File: PYTHON/process_data.py
################################################################################
# Script: Data Preprocessing, Metrics, and Association Analysis
# ******************************************************************************
# Description:
#   This script provides functions for:
#     1. Discretizing numeric variables (equal-width and equal-frequency)
#     2. Computing dataset metrics (entropy, variance, AUC)
#     3. Normalizing and standardizing numeric data
#     4. Filtering variables by metrics
#     5. Computing pairwise association matrices
#     6. Plotting metrics and association matrices
#
# Author: Garikoitz Artola Obando (gartola008@ikasle.ehu.eus)
# Date: 09/11/2025
################################################################################



# LOAD LIBRARIES ---------------------------------------------------------------
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt



# AUXILIAR FUNCTIONS -----------------------------------------------------------
def discretizeEW(x, num_bin):
    x = np.asarray(x)

    if not np.issubdtype(x.dtype, np.number):
        raise ValueError("discretizeEW: x must be numeric")
    if not isinstance(num_bin, int) or num_bin < 1:
        raise ValueError("discretizeEW: num_bin must be integer >= 1")

    if np.all(np.isnan(x)):
        return {"factor": pd.Categorical(x), "cut_points": np.array([])}

    finite_x = x[np.isfinite(x)]
    if len(finite_x) == 0:
        return {"factor": pd.Categorical(x), "cut_points": np.array([])}

    minx, maxx = np.min(finite_x), np.max(finite_x)
    if minx == maxx:
        return {
            "factor": pd.Categorical(["single_value"] * len(x)),
            "cut_points": np.array([])
        }

    breaks = np.linspace(minx, maxx, num_bin + 1)
    cut_points = breaks[1:-1]
    intervals = [-np.inf] + cut_points.tolist() + [np.inf]

    cats = pd.cut(x, bins=intervals, include_lowest=True, right=False)
    return {"factor": cats, "cut_points": cut_points}


def discretizeEF(x, num_bin):
    x = np.asarray(x)

    if not np.issubdtype(x.dtype, np.number):
        raise ValueError("discretizeEF: x must be numeric")
    if not isinstance(num_bin, int) or num_bin < 1:
        raise ValueError("discretizeEF: num_bin must be integer >= 1")

    if np.all(np.isnan(x)):
        return {"factor": pd.Categorical(x), "cut_points": np.array([])}

    finite_mask = ~np.isnan(x)
    n = finite_mask.sum()
    if n == 0:
        return {"factor": pd.Categorical(x), "cut_points": np.array([])}

    num_per_bin = n // num_bin
    rest = n % num_bin

    bin_sizes = [num_per_bin + 1] * rest + [num_per_bin] * (num_bin - rest)

    sorted_idx = np.argsort(np.where(np.isnan(x), np.inf, x))
    sorted_x = x[sorted_idx]

    cum = np.cumsum(bin_sizes)
    cut_indices = cum[:-1]

    cut_points = np.unique(sorted_x[cut_indices])
    intervals = [-np.inf] + cut_points.tolist() + [np.inf]

    cats = pd.cut(x, bins=intervals, include_lowest=True, right=False)
    return {"factor": cats, "cut_points": cut_points}


def entropy(x, base=2):
    x = pd.Series(x).dropna()
    if len(x) == 0:
        return 0.0
    probs = x.value_counts(normalize=True)
    return float(-(probs * np.log(probs) / np.log(base)).sum())


def joint_entropy(x, y, base=2):
    df = pd.DataFrame({"x": x, "y": y}).dropna()
    if df.empty:
        return 0.0
    probs = df.groupby(["x", "y"]).size() / len(df)
    return float(-(probs * np.log(probs) / np.log(base)).sum())


def mutual_information(x, y, base=2):
    return entropy(x, base) + entropy(y, base) - joint_entropy(x, y, base)


def compute_auc_numeric(x, class_col):
    # Convert class to boolean (0/1) if necessary
    class_col = pd.Series(class_col)
    if class_col.dtype.kind in "ifc":  # numeric or categorical
        class_col = class_col.astype("category").cat.codes
    class_col = class_col.astype(bool)

    x = pd.Series(x)

    # Remove NAs
    mask = ~(x.isna() | class_col.isna())
    x = x[mask].values
    class_col = class_col[mask].values

    P = np.sum(class_col)         # total positives
    N = np.sum(~class_col)        # total negatives

    if P == 0 or N == 0:
        raise ValueError("compute_auc_numeric: class must have both True and False")

    # All unique cutpoints
    cuts = np.unique(x)
    TPR = np.zeros(len(cuts))
    FPR = np.zeros(len(cuts))

    for i, cutoff in enumerate(cuts):
        pred = x >= cutoff

        TP = np.sum(pred & class_col)
        FP = np.sum(pred & ~class_col)
        FN = np.sum(~pred & class_col)
        TN = np.sum(~pred & ~class_col)

        TPR[i] = TP / (TP + FN)
        FPR[i] = FP / (FP + TN)

    # Add starting and ending points
    FPR = np.concatenate(([0], FPR, [1]))
    TPR = np.concatenate(([0], TPR, [1]))

    # Trapezoidal rule
    auc = np.sum(np.diff(FPR) * (TPR[:-1] + TPR[1:]) / 2)

    return auc



# FUNCTIONS --------------------------------------------------------------------


# 1. DISCRETIZATION ------------------------------------------------------------
def dis_atribute(attribute, num_bin, mode=False):
    if not np.issubdtype(np.asarray(attribute).dtype, np.number):
        raise ValueError("attribute must be numeric")
    if not isinstance(num_bin, int):
        raise ValueError("num_bin must be int")
    if not isinstance(mode, bool):
        raise ValueError("mode must be bool")

    return discretizeEW(attribute, num_bin) if not mode else discretizeEF(attribute, num_bin)


def dis_dataset(dataset, num_bin, mode=False):
    df = dataset.copy()
    num_cols = df.select_dtypes(include=np.number).columns

    for col in num_cols:
        d = dis_atribute(df[col].values, num_bin, mode)
        df[col] = d["factor"]

    return df



# 2. METRICS -------------------------------------------------------------------
def metrics_dataset(dataset, class_var=None, verbose=False):
    df = dataset.copy()
    if class_var is not None and class_var not in df.columns:
        raise ValueError("class_var not found")

    class_vec = df[class_var] if class_var else None

    results = []

    for col in df.columns:
        if col == class_var:
            continue

        series = df[col]

        if pd.api.types.is_numeric_dtype(series):
            varv = series.var(skipna=True)
            ent = np.nan
            aucv = np.nan

            if class_vec is not None:
                try:
                    aucv = compute_auc_numeric(series, class_vec)
                except Exception:
                    if verbose:
                        print("AUC failed for", col)

            results.append([col, "numeric", varv, ent, aucv])

        else:
            varv = np.nan
            ent = entropy(series)
            aucv = np.nan
            results.append([col, "categorical", varv, ent, aucv])

    return pd.DataFrame(results, columns=["variable", "type", "variance", "entropy", "auc"])



# 3. NORMALIZATION AND STANDARDIZATION -----------------------------------------
def normalize_vector(v):
    v = np.asarray(v)
    finite = v[np.isfinite(v)]
    if len(finite) == 0:
        return np.full(len(v), np.nan)
    minv, maxv = finite.min(), finite.max()
    if minv == maxv:
        return np.zeros(len(v))
    return (v - minv) / (maxv - minv)


def standardize_vector(v):
    v = np.asarray(v)
    finite = v[np.isfinite(v)]
    if len(finite) == 0:
        return np.full(len(v), np.nan)
    mean, sd = finite.mean(), finite.std()
    if sd == 0:
        return np.zeros(len(v))
    return (v - mean) / sd


def normalize_dataset(df):
    out = df.copy()
    for col in out.select_dtypes(include=np.number).columns:
        out[col] = normalize_vector(out[col])
    return out


def standardize_dataset(df):
    out = df.copy()
    for col in out.select_dtypes(include=np.number).columns:
        out[col] = standardize_vector(out[col])
    return out



# 4. FILTERING -----------------------------------------------------------------
def filter_by_metric(dataset, metric, threshold, class_var=None):
    if metric not in ["entropy", "variance", "auc"]:
        raise ValueError("metric must be 'entropy', 'variance', or 'auc'")

    met = metrics_dataset(dataset, class_var)

    if metric == "entropy":
        keep = met.loc[met["entropy"] >= threshold, "variable"]
    elif metric == "variance":
        keep = met.loc[met["variance"] >= threshold, "variable"]
    else:
        if class_var is None:
            raise ValueError("class_var required for AUC")
        keep = met.loc[met["auc"] >= threshold, "variable"]

    final_vars = [v for v in keep if v in dataset.columns]
    if class_var and class_var not in final_vars:
        final_vars.append(class_var)

    return dataset[final_vars]



# 5. PAIRWISE ASSOCIATION MATRIX -----------------------------------------------
def pairwise_assoc_matrix(dataset, num_bin=10, normalize_mi=True, mode=False):

    df = dataset.copy()
    cols = df.columns
    n = len(cols)
    mat = pd.DataFrame(np.zeros((n, n)), index=cols, columns=cols)

    # Discretize numeric columns once
    df_discretized = dis_dataset(df, num_bin=num_bin, mode=mode)

    for i in cols:
        for j in cols:
            if i == j:
                mat.loc[i, j] = 1
                continue

            a, b = df[i], df[j]

            # Both numeric → Pearson correlation
            if pd.api.types.is_numeric_dtype(a) and pd.api.types.is_numeric_dtype(b):
                mat.loc[i, j] = a.corr(b)
            else:
                # Use discretized version for mutual information
                A, B = df_discretized[i], df_discretized[j]

                mi = mutual_information(A, B)  # Assumes you have this function

                if normalize_mi:
                    Ha = entropy(A)  # Assumes you have this function
                    Hb = entropy(B)
                    denom = np.sqrt(Ha * Hb)
                    mat.loc[i, j] = mi / denom if denom > 0 else 0
                else:
                    mat.loc[i, j] = mi

    return mat



# 6. PLOTS ---------------------------------------------------------------------
def plot_auc(metrics_df):
    df = metrics_df.dropna(subset=["auc"])

    variables = df["variable"]
    aucs = df["auc"]

    plt.bar(variables, aucs)
    plt.xticks(rotation=90)
    plt.ylim(0, 1)
    plt.ylabel("AUC")
    plt.title("AUC per numeric attribute")
    plt.axhline(0.5, linestyle="--")
    plt.show()


def plot_assoc_matrix(mat):
    plt.imshow(mat, cmap="bwr", vmin=-1, vmax=1)
    plt.colorbar()
    plt.xticks(range(mat.shape[1]), mat.columns, rotation=90)
    plt.yticks(range(mat.shape[0]), mat.index)
    plt.title("Association matrix")
    plt.show()

File: R/process_data.R
################################################################################
# Script: Data Preprocessing, Metrics, and Association Analysis
# ******************************************************************************
# Descripción:
#   Este script proporciona funciones para:
#     1. Discretizar variables numéricas (intervalos iguales y frecuencias iguales)
#     2. Calcular métricas del conjunto de datos (entropía, varianza, AUC)
#     3. Normalizar y estandarizar datos numéricos
#     4. Filtrar variables según métricas
#     5. Calcular matrices de asociación por pares
#     6. Representar gráficamente métricas y matrices de asociación
#
# Autor: Garikoitz Artola Obando (gartola008@ikasle.ehu.eus)
# Fecha: 09/11/2025
################################################################################


# LOAD LIBRARIES ---------------------------------------------------------------
library(ggplot2)
library(reshape2)



# AUXILIAR FUNCTIONS -----------------------------------------------------------
discretizeEW <- function(x, num.bins) {
  if (!is.numeric(x)) stop("discretizeEW: x must be numeric")
  if (!is.numeric(num.bins) || length(num.bins) != 1) stop("discretizeEW: num.bins must be single numeric")
  if (num.bins < 1) stop("discretizeEW: num.bins must be >= 1")
  if (all(is.na(x))) return(list(factor = factor(x), cut.points = numeric(0)))
  finite.x <- x[is.finite(x)]
  if (length(finite.x) == 0) return(list(factor = factor(x), cut.points = numeric(0)))
  minx <- min(finite.x)
  maxx <- max(finite.x)

  if (minx == maxx) { # all same value
    cut.points <- numeric(0)
    x.discretized <- factor(rep("single_value", length(x)), levels = "single_value")
    return(list(factor = x.discretized, cut.points = cut.points))
  }

  # create breakpoints (num.bins equal width)
  breaks <- seq(from = minx, to = maxx, length.out = num.bins + 1)
  # internal cut points excludes endpoints
  cut.points <- sort(unique(breaks))[ - c(1, length(breaks)) ]
  intervalos <- c(-Inf, cut.points, Inf)
  x.discretized <- cut(x, breaks = intervalos, include.lowest = TRUE, right = FALSE)
  return(list(factor = x.discretized, cut.points = cut.points))
}


discretizeEF <- function(x, num.bins) {
  if (!is.numeric(x)) stop("discretizeEF: x must be numeric")
  if (!is.numeric(num.bins) || length(num.bins) != 1) stop("discretizeEF: num.bins must be single numeric")
  if (num.bins < 1) stop("discretizeEF: num.bins must be >= 1")
  if (all(is.na(x))) return(list(factor = factor(x), cut.points = numeric(0)))
  n <- sum(!is.na(x))
  if (n == 0) return(list(factor = factor(x), cut.points = numeric(0)))

  # compute number per bin (some bins may have one additional element)
  num.perBin <- n %/% num.bins
  rest.perBin <- n %% num.bins

  # sizes distribution from smallest to largest is fine but we will build cutpoints from sorted x
  bin.sizes <- c(rep(num.perBin + 1, rest.perBin), rep(num.perBin, num.bins - rest.perBin))
  sorted_idx <- order(x, na.last = TRUE)
  sorted_x <- x[sorted_idx]

  # cumulative sizes to pick cut values (end of each bin except last)
  cum.sizes <- cumsum(bin.sizes)
  cut.indices <- cum.sizes[1:(length(cum.sizes) - 1)]
  # if there are ties that span cut indices, it's okay: we pick the observed value at that index
  cut.points <- unique(sorted_x[cut.indices])
  intervalos <- c(-Inf, cut.points, Inf)
  x.discretized <- cut(x, breaks = intervalos, include.lowest = TRUE, right = FALSE)
  return(list(factor = x.discretized, cut.points = cut.points))
}


entropy <- function(x, base = 2, na.rm = TRUE) {
  if (na.rm) x <- x[!is.na(x)]
  if (length(x) == 0) return(0)
  probs <- table(x) / length(x)
  probs <- probs[probs > 0]
  H <- -sum(probs * log(probs, base = base))
  return(as.numeric(H))
}


# joint entropy for two discrete variables
joint_entropy <- function(x, y, base = 2, na.rm = TRUE) {
  if (na.rm) {
    ok <- !(is.na(x) | is.na(y))
    x <- x[ok]; y <- y[ok]
  }
  if (length(x) == 0) return(0)
  tab <- table(x, y)
  probs <- tab / sum(tab)
  probs <- probs[probs > 0]
  H <- -sum(probs * log(probs, base = base))
  return(as.numeric(H))
}


# mutual information I(X;Y) = H(X)+H(Y)-H(X,Y)
mutual_information <- function(x, y, base = 2, na.rm = TRUE) {
  Hx <- entropy(x, base = base, na.rm = na.rm)
  Hy <- entropy(y, base = base, na.rm = na.rm)
  Hxy <- joint_entropy(x, y, base = base, na.rm = na.rm)
  MI <- Hx + Hy - Hxy
  return(as.numeric(MI))
}


compute_auc_numeric <- function(x, class) {
  # class must be TRUE/FALSE or convert to logical
  if (is.numeric(class) || is.factor(class))
    class <- as.logical(as.numeric(as.factor(class)) - 1)

  # Remove NA
  ok <- !(is.na(x) | is.na(class))
  x <- x[ok]
  class <- class[ok]

  # Sort by predictor value
  o <- order(x)
  x <- x[o]
  class <- class[o]

  # total positives / negatives
  P <- sum(class)
  N <- sum(!class)

  if (P == 0 || N == 0)
    stop("compute_auc_numeric: class must have both TRUE and FALSE")

  # All possible cutpoints = each unique value
  cuts <- unique(x)
  TPR <- numeric(length(cuts))
  FPR <- numeric(length(cuts))
  for (i in seq_along(cuts)) {
    cutoff <- cuts[i]

    # prediction: x >= cutoff is TRUE
    pred <- x >= cutoff

    TP <- sum(pred &  class)
    FP <- sum(pred & !class)
    FN <- sum(!pred &  class)
    TN <- sum(!pred & !class)

    TPR[i] <- TP / (TP + FN)
    FPR[i] <- FP / (FP + TN)
  }

  # Add starting point (0,0) and ending point (1,1)
  FPR <- c(0, FPR, 1)
  TPR <- c(0, TPR, 1)

  # Compute area using trapezoids
  auc <- sum( diff(FPR) * (head(TPR, -1) + tail(TPR, -1)) / 2 )
  return(auc)
}



# FUNCTIONS --------------------------------------------------------------------

# 1. DISCRETIZATION ------------------------------------------------------------
dis.attribute <- function(attribute, num.bin, mode = FALSE) {
  if (!is.numeric(attribute)) stop("Error: attribute is not numeric")
  if (!is.numeric(num.bin) || length(num.bin) != 1) stop("Error: num.bin must be a single numeric")
  if (!is.logical(mode) || length(mode) != 1) stop("Error: mode must be boolean (logical)")
  if (mode == FALSE) {
    return(discretizeEW(attribute, num.bin))
  } else {
    return(discretizeEF(attribute, num.bin))
  }
}


dis.dataset <- function(dataset, num.bin, mode = FALSE) {
  # --- Error checking ---
  if (!is.data.frame(dataset)) stop("Error: dataset must be a data.frame")
  if (!is.numeric(num.bin) || length(num.bin) != 1) stop("Error: num.bin must be a single numeric")
  if (!is.logical(mode) || length(mode) != 1) stop("Error: mode must be boolean (logical)")

  # Copy to avoid modifying in place
  res.df <- dataset
  # Select numeric columns
  numeric.cols <- sapply(dataset, is.numeric)

  # Loop over numeric columns and discretize
  for (colname in names(dataset)[numeric.cols]) {
    if (mode == FALSE) {
      d <- discretizeEW(dataset[[colname]], num.bin)
    } else {
      d <- discretizeEF(dataset[[colname]], num.bin)
    }
    res.df[[colname]] <- d$factor   # replace column with discretized factor
  }
  return(res.df)
}



# 2. METRICS -------------------------------------------------------------------
metrics_dataset <- function(dataset, class.var = NULL, discretize_for_auc = FALSE, bins = 10, verbose = FALSE) {
  if (!is.data.frame(dataset)) stop("dataset must be a data.frame")

  if (!is.null(class.var) && !(class.var %in% names(dataset))) stop("metrics_dataset: class.var not found")

  class.vec <- if (!is.null(class.var)) dataset[[class.var]] else NULL

  result <- data.frame(
    variable = character(0),
    type = character(0),
    variance = numeric(0),
    entropy = numeric(0),
    auc = numeric(0),
    stringsAsFactors = FALSE
  )

  for (nm in names(dataset)) {
    if (!is.null(class.var) && nm == class.var) next

    col <- dataset[[nm]]

    if (is.numeric(col)) {
      varv <- ifelse(all(is.na(col)), NA_real_, var(col, na.rm = TRUE))
      ent <- NA_real_
      aucv <- NA_real_

      if (!is.null(class.vec)) {
        tmp <- try({
          aucv <- compute_auc_numeric(col, class.vec)
        }, silent = TRUE)

        if (inherits(tmp, "try-error") && verbose)
          message("AUC failed for ", nm)
      }

      typ <- "numeric"

    } else {
      varv <- NA_real_
      ent <- entropy(col)
      aucv <- NA_real_
      typ <- "categorical"
    }

    result <- rbind(
      result,
      data.frame(variable = nm, type = typ, variance = varv, entropy = ent, auc = aucv)
    )
  }
  rownames(result) <- NULL
  return(result)
}



# 3. NORMALIZATION AND STANDARDIZATION -----------------------------------------
normalize_vector <- function(v, na.rm = TRUE) {
  if (!is.numeric(v)) stop("normalize_vector: v must be numeric")
  if (na.rm) v2 <- v[is.finite(v)] else v2 <- v
  if (length(v2) == 0) return(rep(NA_real_, length(v)))
  minv <- min(v2)
  maxv <- max(v2)
  denom <- (maxv - minv)
  if (denom == 0) return(rep(0, length(v))) # all equal -> zero vector
  out <- (v - minv) / denom
  return(out)
}


standardize_vector <- function(v, na.rm = TRUE) {
  if (!is.numeric(v)) stop("standardize_vector: v must be numeric")
  if (na.rm) v2 <- v[is.finite(v)] else v2 <- v
  if (length(v2) == 0) return(rep(NA_real_, length(v)))
  m <- mean(v2)
  s <- sd(v2)
  if (s == 0) return(rep(0, length(v)))
  out <- (v - m) / s
  return(out)
}


normalize_dataset <- function(df, na.rm = TRUE) {
  if (!is.data.frame(df)) stop("normalize_dataset: df must be data.frame")
  out <- df
  for (nm in names(df)) {
    if (is.numeric(df[[nm]])) out[[nm]] <- normalize_vector(df[[nm]], na.rm = na.rm)
  }
  return(out)
}


standardize_dataset <- function(df, na.rm = TRUE) {
  if (!is.data.frame(df)) stop("standardize_dataset: df must be data.frame")
  out <- df
  for (nm in names(df)) {
    if (is.numeric(df[[nm]])) out[[nm]] <- standardize_vector(df[[nm]], na.rm = na.rm)
  }
  return(out)
}



# 4. FILTERING -----------------------------------------------------------------
filter_by_metric <- function(dataset, metric = c("entropy", "variance", "auc"), threshold, class.var = NULL, bins = 10) {
  metric <- match.arg(metric)
  if (!is.data.frame(dataset)) stop("filter_by_metric: dataset must be a data.frame")
  if (!is.numeric(threshold) || length(threshold) != 1) stop("filter_by_metric: threshold must be single numeric")
  # compute metrics
  met.df <- metrics_dataset(dataset, class.var = class.var)
  keep.vars <- character(0)
  if (metric == "entropy") {
    # entropy available for categorical only
    keep.vars <- met.df$variable[!is.na(met.df$entropy) & met.df$entropy >= threshold]
  } else if (metric == "variance") {
    keep.vars <- met.df$variable[!is.na(met.df$variance) & met.df$variance >= threshold]
  } else if (metric == "auc") {
    if (is.null(class.var)) stop("filter_by_metric: class.var must be provided to compute AUC")
    keep.vars <- met.df$variable[!is.na(met.df$auc) & met.df$auc >= threshold]
  }
  # return dataset that contains class.var (if present) + kept variables
  final.vars <- intersect(c(keep.vars, class.var), names(dataset))
  if (length(final.vars) == 0) {
    warning("filter_by_metric: no variables meet the threshold; returning empty data.frame with zero columns")
    return(dataset[ , FALSE, drop = FALSE])
  }
  return(dataset[, final.vars, drop = FALSE])
}



# 5. PAIRWISE ASSOCIATION MATRIX -----------------------------------------------
pairwise_assoc_matrix <- function(dataset, num.bin = 10, normalize_mi = TRUE, mode = FALSE, na.rm = TRUE) {
  if (!is.data.frame(dataset)) stop("pairwise_assoc_matrix: dataset must be data.frame")

  n <- ncol(dataset)
  names_cols <- names(dataset)
  mat <- matrix(NA_real_, nrow = n, ncol = n, dimnames = list(names_cols, names_cols))

  # Determine column types
  types <- sapply(dataset, function(x) if (is.numeric(x)) "numeric" else "categorical")

  # Discretize numeric columns once using existing functions
  dataset_discretized <- dis.dataset(dataset, num.bin = num.bin, mode = mode)

  # Loop over all pairs
  for (i in seq_len(n)) {
    for (j in seq_len(n)) {
      xi <- dataset[[i]]
      xj <- dataset[[j]]

      if (i == j) {
        mat[i, j] <- 1
        next
      }

      # Numeric-Numeric: Pearson correlation
      if (types[i] == "numeric" && types[j] == "numeric") {
        mat[i, j] <- suppressWarnings(cor(xi, xj, use = "pairwise.complete.obs"))
      } else {
        # At least one categorical: use mutual information on factors
        fi <- if (types[i] == "numeric") dataset_discretized[[i]] else as.factor(xi)
        fj <- if (types[j] == "numeric") dataset_discretized[[j]] else as.factor(xj)

        mi <- mutual_information(fi, fj, na.rm = na.rm)  # your MI function

        if (normalize_mi) {
          Hx <- entropy(fi, na.rm = na.rm)
          Hy <- entropy(fj, na.rm = na.rm)
          denom <- sqrt(Hx * Hy)
          mat[i, j] <- if (denom == 0) 0 else mi / denom
        } else {
          mat[i, j] <- mi
        }
      }
    }
  }
  return(mat)
}



# 6. PLOTS ---------------------------------------------------------------------
plot_auc <- function(metrics_df) {
  # --- checks ---
  if (!("variable" %in% names(metrics_df)) || !("auc" %in% names(metrics_df))) {
    stop("plot_auc: metrics_df must contain columns 'variable' and 'auc'")
  }

  aucs <- metrics_df$auc
  vars <- metrics_df$variable
  valid <- !is.na(aucs)

  if (sum(valid) == 0) {
    warning("plot_auc: no AUC values to plot")
    return(NULL)
  }

  aucs <- aucs[valid]
  vars <- vars[valid]

  # Sort by descending AUC
  ord <- order(aucs, decreasing = TRUE)
  aucs <- aucs[ord]
  vars <- vars[ord]

  # --- Plot ---
  bar_positions <- barplot(
    height = aucs,
    names.arg = vars,
    las = 2,
    ylim = c(0, 1),
    col = "skyblue",
    border = "gray40",
    main = "AUC per numeric attribute",
    ylab = "AUC"
  )

  abline(h = 0.5, col = "gray50", lty = 2)

  invisible(NULL)
}


plot_assoc_matrix <- function(mat, main = "Association matrix") {
  if (!is.matrix(mat))
    stop("plot_assoc_matrix: mat must be a matrix")

  # Convert matrix to long format for ggplot
  df <- melt(mat)
  colnames(df) <- c("Row", "Col", "Value")

  # Reverse y-axis order (so top matches your image())
  df$Row <- factor(df$Row, levels = rev(rownames(mat)))
  df$Col <- factor(df$Col, levels = colnames(mat))

  # bwr palette (blue → white → red)
  palette_colors <- colorRampPalette(c("blue", "white", "red"))(200)

  ggplot(df, aes(x = Col, y = Row, fill = Value)) +
    geom_tile(color = NA) +
    scale_fill_gradientn(
      colours = palette_colors,
      limits = c(-1, 1),
      name = "Value"
    ) +
    labs(title = main, x = "", y = "") +
    theme_minimal(base_size = 14) +
    theme(
      axis.text.x  = element_text(angle = 90, vjust = 0.5, hjust = 1),
      panel.grid   = element_blank(),
      plot.title   = element_text(hjust = 0.5)
    )
}



